File: geekbang/algorithm-models/6.1- Trie 树代码模板/Trie.cpp
//C/C++
class Trie {
public:
    /** Initialize your data structure here. */
    Trie() {
        root = new Node();
    }


    /** Inserts a word into the trie. */
    void insert(string word) {
        find(word, true, true);
    }


    /** Returns if the word is in the trie. */
    bool search(string word) {
        return find(word, true, false);
    }


    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        return find(prefix, false, false);
    }


private:
    struct Node {
        int count;
        unordered_map<char, Node*> child;
        Node(): count(0) {}
    };
    Node* root;


    bool find(const string& s, bool exact_match, bool insert_if_not_exist) {
        Node* curr = root;
        for (char c : s) {
            if (curr->child.find(c) == curr->child.end()) {
                if (!insert_if_not_exist) return false;
                curr->child[c] = new Node();
            }
            curr = curr->child[c];
        }
        if (insert_if_not_exist) curr->count++;
        return exact_match ? curr->count > 0 : true;
    }
};

File: geekbang/algorithm-models/6.1- Trie 树代码模板/Trie.java
//Java
import java.util.HashMap;
class Trie {
    /** Initialize your data structure here. */
    public Trie() {
        root = new Node();
    }


    /** Inserts a word into the trie. */
    public void insert(String word) {
        find(word, true, true);
    }


    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        return find(word, true, false);
    }


    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        return find(prefix, false, false);
    }


    class Node {
        public int count;
        public HashMap<Character, Node> child;
        public Node() { count = 0; child = new HashMap<>(); }
    }
    Node root;


    boolean find(String s, boolean exact_match, boolean insert_if_not_exist) {
        Node curr = root;
        for (Character c : s.toCharArray()) {
            if (!curr.child.containsKey(c)) {
                if (!insert_if_not_exist) return false;
                curr.child.put(c, new Node());
            }
            curr = curr.child.get(c);
        }
        if (insert_if_not_exist) curr.count++;
        return exact_match ? curr.count > 0 : true;
    }
}
